Packet++/header/SomeIpSdLayer.h

```diff
@@ -0,0 +1,998 @@
+		uint16_t eventGroupID;
+	};
+#pragma pack(pop)
+
+	SomeIpSdEntryType2(SomeIpSdLayer *pSomeIpSdLayer, size_t offset) : SomeIpSdEntry(pSomeIpSdLayer, offset) {}
+
+	SomeIpSdEntryType2 *doCloneTo(SomeIpSdLayer *pSomeIpSdLayer, size_t offset) const override;
+
+	std::unique_ptr<SomeIpSdEntryType2> cloneTo(SomeIpSdLayer *pSomeIpSdLayer, size_t offset) const
+	{
+		return std::unique_ptr<SomeIpSdEntryType2>(doCloneTo(pSomeIpSdLayer, offset));
+	}
+
+	SomeIpSdEntryType2 &operator=(const SomeIpSdEntryType2 &rhs)
+	{
+		SomeIpSdEntry::operator=(static_cast<const SomeIpSdEntry &>(rhs));
+		return *this;
+	}
+};
+
+typedef std::unique_ptr<SomeIpSdEntry> EntryPtr;
+typedef std::vector<EntryPtr> EntriesVec;
+
+/**
+ * @class SomeIpSdLayer
+ * Implementation of the SOME/IP-SD protocol
+ */
+class SomeIpSdLayer : public SomeIpLayer
+{
+public:
+	friend class SomeIpSdEntry;
+
+	/**
+	 * A constructor that creates the layer from an existing packet raw data
+	 * @param[in] data A pointer to the raw data
+	 * @param[in] dataLen Size of the data in bytes
+	 * @param[in] prevLayer A pointer to the previous layer
+	 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+	 */
+	SomeIpSdLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet);
+
+	/**
+	 * Construct a new SomeIpSdLayer object
+	 * @param[in] serviceID Service ID
+	 * @param[in] methodID Method ID
+	 * @param[in] clientID Client ID
+	 * @param[in] sessionID Session ID
+	 * @param[in] interfaceVersion Interface Version
+	 * @param[in] type Type of the message
+	 * @param[in] returnCode Return Code
+	 * @param[in] flags Flags that shall be used in the header
+	 */
+	SomeIpSdLayer(uint16_t serviceID, uint16_t methodID, uint16_t clientID, uint16_t sessionID,
+				  uint8_t interfaceVersion, MsgType type, uint8_t returnCode, uint8_t flags);
+
+	/**
+	 * Destroy the layer object
+	 */
+	~SomeIpSdLayer() {}
+
+	/**
+	 * Checks if given port is a SomeIP SD protocol port
+	 * @param[in] port Port to check
+	 * @return true if SomeIP protocol port, false if not
+	 */
+	static bool isSomeIpSdPort(uint16_t port) { return port == 30490; }
+
+	/**
+	 * Get the Entries contained in the layer
+	 * @return std::vector<SomeIpSdEntry *>
+	 */
+	std::vector<SomeIpSdEntry *> getEntries() const;
+
+	/**
+	 * Get the Flags of the layer
+	 * @return uint8_t Flags
+	 */
+	uint8_t getFlags() const;
+
+	/**
+	 * Set the Flags of the layer
+	 * @param[in] flags Flags to set
+	 */
+	void setFlags(uint8_t flags);
+
+	/**
+	 * Adds a given entry to the layer and returns the pointer to the added entry.
+	 * @tparam TEntryPtr Template for ptr to derived entries.
+	 * @param[in] pEntry Pointer to the entry that shall be added to the layer
+	 * @return TEntryPtr& Returns a pointer to the entry in the layer. Note that this pointer will be invalid as
+	 * soon as an further option or entry is added to the layer.
+	 */
+	template <class TEntryPtr> TEntryPtr &addEntry(const TEntryPtr &pEntry);

```

Comment:
I suggest this method doesn't accept a pointer but a const reference (`const SomeIpSdEntry&`) and just copies its data to the packet